File: demo/src/main/java/com/mta/tehreer/demo/TextViewWidgetActivity.java
/*
 * Copyright (C) 2021 Muhammad Tayyab Akram
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.mta.tehreer.demo;

import android.content.res.AssetManager;
import android.graphics.Color;
import android.os.Bundle;
import android.text.SpannableStringBuilder;
import android.text.Spanned;
import android.text.style.ForegroundColorSpan;
import android.view.Gravity;
import android.widget.SeekBar;

import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;

import com.mta.tehreer.graphics.TypefaceManager;
import com.mta.tehreer.widget.TTextView;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.Reader;

public class TextViewWidgetActivity extends AppCompatActivity {
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_text_view_widget);

        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
        }

        final TTextView textView = findViewById(R.id.text_view);
        textView.setTypeface(TypefaceManager.getTypeface(R.id.typeface_noorehuda));
        textView.setGravity(Gravity.CENTER_HORIZONTAL);
        textView.setSpanned(parseSurah());
        textView.setLineHeightMultiplier(0.80f);
        textView.setJustificationEnabled(true);

        final SeekBar justificationLevelBar = findViewById(R.id.seek_bar_justification_level);
        justificationLevelBar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int i, boolean b) {
                textView.setJustificationLevel(((i / 4.0f) * 0.35f) + 0.65f);
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) { }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) { }
        });
    }

    @Override
    public boolean onSupportNavigateUp() {
        onBackPressed();
        return true;
    }

    private Spanned parseSurah() {
        try {
            final AssetManager assetManager = getAssets();
            final InputStream stream = assetManager.open("AlKahf.json");

            final char[] buffer = new char[1024];
            final StringBuilder out = new StringBuilder();
            final Reader in = new InputStreamReader(stream);
            int length;

            while ((length = in.read(buffer, 0, buffer.length)) > 0) {
                out.append(buffer, 0, length);
            }

            final String jsonString = out.toString();
            final SpannableStringBuilder surah = new SpannableStringBuilder();

            final JSONArray ayahsJson = new JSONArray(jsonString);
            final int ayahCount = ayahsJson.length();

            for (int i = 0; i < ayahCount; i++) {
                final JSONObject ayahJson = ayahsJson.getJSONObject(i);
                final String ayahText = ayahJson.getString("text");
                final JSONArray attrsJson = ayahJson.getJSONArray("attributes");
                final int attrCount = attrsJson.length();

                final int offset = surah.length();
                surah.append(ayahText);

                for (int j = 0; j < attrCount; j++) {
                    final JSONObject attrJson = attrsJson.getJSONObject(j);
                    final int start = attrJson.getInt("start");
                    final int end = attrJson.getInt("end");
                    final String colorString = attrJson.getString("color");
                    final int color = Color.parseColor("#" + colorString);

                    surah.setSpan(new ForegroundColorSpan(color), start + offset,
                                  end + offset, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
                }

                surah.append(i == 0 ? "\n" : "  ");
            }

            return surah;
        } catch (IOException e) {
            e.printStackTrace();
        } catch (JSONException e) {
            e.printStackTrace();
        }

        return null;
    }
}

File: demo/src/main/java/com/mta/tehreer/demo/TextViewWidgetActivity.kt
/*
 * Copyright (C) 2023 Muhammad Tayyab Akram
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.mta.tehreer.demo

import android.graphics.Color
import android.os.Bundle
import android.text.SpannableStringBuilder
import android.text.Spanned
import android.text.style.ForegroundColorSpan
import android.util.TypedValue
import android.view.Gravity
import android.widget.SeekBar
import androidx.appcompat.app.AppCompatActivity
import com.mta.tehreer.graphics.TypefaceManager
import org.json.JSONArray
import org.json.JSONException
import java.io.IOException
import java.io.InputStreamReader
import java.io.Reader

private const val MIN_TEXT_SIZE = 20f
private const val MAX_TEXT_SIZE = 56f

class TextViewWidgetActivity : AppCompatActivity() {
    private lateinit var textView: QuranTextView
    private lateinit var textSizeBar: SeekBar

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_text_view_widget)

        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        textSizeBar = findViewById(R.id.seek_bar_text_size)
        textSizeBar.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar, i: Int, b: Boolean) { }

            override fun onStartTrackingTouch(seekBar: SeekBar) {
                clearAyahHighlighting()
            }

            override fun onStopTrackingTouch(seekBar: SeekBar) {
                updateTextSize()
            }
        })

        textView = findViewById(R.id.text_view)
        textView.apply {
            updateTextSize()
            setGravity(Gravity.CENTER_HORIZONTAL)
            typeface = TypefaceManager.getTypeface(R.id.typeface_noorehuda)
            spanned = parseSurah()
            lineHeightMultiplier = 0.80f
            isJustificationEnabled = true
            separatorColor = Color.GRAY
            highlightingColor = resources.getColor(R.color.colorHighlight)
        }
    }

    override fun onSupportNavigateUp(): Boolean {
        onBackPressed()
        return true
    }

    private fun parseSurah(): Spanned? {
        try {
            val stream = assets.open("AlKahf.json")
            val buffer = CharArray(1024)
            val out = StringBuilder()
            val `in`: Reader = InputStreamReader(stream)
            var length: Int

            while (`in`.read(buffer, 0, buffer.size).also { length = it } > 0) {
                out.append(buffer, 0, length)
            }

            val jsonString = out.toString()
            val surah = SpannableStringBuilder()

            val ayahsJson = JSONArray(jsonString)
            val ayahCount = ayahsJson.length()

            for (i in 0 until ayahCount) {
                val ayahJson = ayahsJson.getJSONObject(i)
                val ayahText = ayahJson.getString("text")
                val attrsJson = ayahJson.getJSONArray("attributes")
                val attrCount = attrsJson.length()

                val offset = surah.length
                surah.append(ayahText)
                surah.setSpan(
                    QuranTextView.AyahSpan(),
                    offset, surah.length,
                    Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
                )

                for (j in 0 until attrCount) {
                    val attrJson = attrsJson.getJSONObject(j)
                    val start = attrJson.getInt("start")
                    val end = attrJson.getInt("end")
                    val colorString = attrJson.getString("color")
                    val color = Color.parseColor("#$colorString")

                    surah.setSpan(
                        ForegroundColorSpan(color),
                        start + offset,
                        end + offset,
                        Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
                    )
                }

                surah.append(if (i == 0) "\n" else "  ")
            }

            return surah
        } catch (e: IOException) {
            e.printStackTrace()
        } catch (e: JSONException) {
            e.printStackTrace()
        }

        return null
    }

    private fun spToPx(sp: Float) =
        TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_SP, sp, resources.displayMetrics)

    private fun clearAyahHighlighting() {
        textView.clearAyahHighlighting()
    }

    private fun updateTextSize() {
        val ratio = (textSizeBar.progress.toFloat() / textSizeBar.max.toFloat())
        val multiplier = MAX_TEXT_SIZE - MIN_TEXT_SIZE
        val sp = (ratio * multiplier) + MIN_TEXT_SIZE

        textView.textSize = spToPx(sp)
    }
}
